### ISSUE TITLE
Weighted mean in cube.aggregated_by

### ISSUE THREAD

**Issue by ledm** [SciTools/iris#3290]
Hi,

is it possible to use weights to calculate the mean while using `cube.aggregated_by`.

For instance, I have a cube with several years of monthly data, and I want to produce an annual mean. I have previously used:
```
cube.aggregated_by('year', iris.analysis.MEAN)
```
However, this means that the shorter months have the same weighting as the longer months. Is there a weight to weight this calculation by some given weights. (Lets assume that I've calculated them elsewhere!)
ie:

```
cube.aggregated_by('year', iris.analysis.MEAN, weights=time_weights)
```

Thanks!

**Comment by pp-mo** [comment]
Absolutely yes, it does supporta 'weights' keyword
see [here](https://scitools.org.uk/iris/docs/latest/iris/iris/analysis.html#iris.analysis.MEAN)

**Comment by ledm** [comment]
Hi Patrick!

I'm seeing the error message:
```
 ValueError: Invalid Aggregation, aggregated_by() cannot use weights.
```
Which seems to come from:
```
            # We can't handle weights
            if isinstance(aggregator, iris.analysis.WeightedAggregator) and \
                    aggregator.uses_weighting(**kwargs):
>               raise ValueError('Invalid Aggregation, aggregated_by() cannot use'
                                 ' weights.')
```
in iris/cube.py:3363 and we're in iris version 2.2.0.

It probably isn't the culprit, but the function that you're linking is the `iris.analysis.MEAN` function, not the `aggregated_by` function that I'm using. Is there perhaps a problem with the way I'm passing the weights to `aggregated_by`?

**Comment by pp-mo** [comment]
Sorry, my bad !! :-1: 
I was confusing the functions of the aggregator with the stats routine (aggregated_by instead of collapsed).

**Comment by bjlittle** [comment]
In principle, this _seems_ possible...

I think it might be worthwhile taking a step back here and holistically rationalising how aggregation and collapse works. Unifying behaviour for use of weights (where appropriate for the statistic) consistently would be a good win - plus there is some technical debt for us to purge, as there is a major overlap between aggregation and collapse that we should take advantage of; I've been itching to do that for years. 

The topic of operator laziness (#3280) is also relevant here... and that needs a wee bit more thought. You'd certainly throw away coordinate laziness, but it _seems_ reasonable to expect to keep the data lazy.

**Comment by pp-mo** [comment]
> Unifying behaviour for use of weights ... consistently
> operator laziness ... throw away coordinate laziness, but ... keep the data lazy

:+1: great summary!

**Comment by rcomer** [comment]
This was implemented in #4589

**Comment by ledm** [comment]
Great work, thanks to the iris team! This will be very useful in the future. Now if I can only remember what I was trying to do back in March 2019, lol.
